core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,443 @@
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	simulator BundleSimulator
+
+	bundleReceiverClients map[string]*rpc.Client
+}
+
+func New(config Config, mevConfig miner.MevConfig) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:                config,
+		mevConfig:             mevConfig,
+		bundles:               make(map[common.Hash]*types.Bundle),
+		bundleHeap:            make(BundleHeap, 0),
+		bundleReceiverClients: make(map[string]*rpc.Client),
+	}
+	if pool.mevConfig.MevReceivers == nil {
+		return pool
+	}
+	for _, v := range mevConfig.MevReceivers {
+		pool.register(v)
+	}
+	if len(pool.bundleReceiverClients) == 0 {
+		log.Error("No valid bundleReceivers")
+	}
+
+	return pool
+}
+
+func (p *BundlePool) register(url string) {
+	var receiverClient *rpc.Client
+	if url != "" {
+		var err error
+		for i := 1; i < dialAttempts+1; i++ {
+			receiverClient, err = rpc.Dial(url)
+			if err != nil {
+				log.Error("failed to dial bundle receiver", "attempts", i, "url", url, "err", err)
+				if i == dialAttempts {
+					return
+				}
+				time.Sleep(dialSleep)
+				continue
+			}
+			p.bundleReceiverClients[url] = receiverClient
+			log.Info("success to dial bundle receiver", "url", url)
+			return
+		}
+	}
+}
+
+func (p *BundlePool) SetBundleSimulator(simulator BundleSimulator) {
+	p.simulator = simulator
+}
+
+func (p *BundlePool) Init(gasTip *big.Int, head *types.Header, reserve txpool.AddressReserver) error {
+	return nil
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle, originBundle *types.SendBundleArgs) error {
+	if p.simulator == nil {
+		return ErrSimulatorMissing
+	}
+
+	if bundle.MinTimestamp > uint64(time.Now().Unix()+maxMinTimestampFromNow) {
+		return ErrBundleTimestampTooHigh
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+
+	if price.Cmp(p.minimalBundleGasPrice()) < 0 && p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		return ErrBundleGasPriceLow
+	}
+	bundle.Price = price
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	hash := bundle.Hash()
+	if _, ok := p.bundles[hash]; ok {
+		return ErrBundleAlreadyExist
+	}
+
+	for url, cli := range p.bundleReceiverClients {
+		cli := cli
+		url := url
+		go func() {
+			var hash common.Hash
+			err := cli.CallContext(context.Background(), &hash, "eth_sendBundle", *originBundle)
+			if err != nil {
+				bundleDeliverFailed.Inc(1)
+				log.Error("failed to deliver bundle to receiver", "url", url, "err", err)
+			}
+		}()
+		bundleDeliverAll.Inc(1)
+	}
+
+	for p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		p.drop()
```

Comment:
I think every time when we dropped a bundle, we should compare the price again. Otherwise, a huge bundle whose price is only higher the the minimal price bundle will result in other bundles dropped.